File: scraper.rb
require 'nokogiri'
require 'httparty'
require 'byebug'
require 'json'


def scraper
    url = "https://www.yr.no/en/forecast/daily-table/2-2179537/New%20Zealand/Wellington/Wellington%20City/Wellington"
    unparsed_page = HTTParty.get(url)
    parsed_page = Nokogiri::HTML(unparsed_page)
    weatherForecast = parsed_page.css('div.forecast-page') 
    currentWeather = weatherForecast.css('div.now-hero__slide')
    @currentTime = weatherForecast.css('h3.now-hero__heading').text.gsub("Hour", "")
    @currentWeatherIcon = currentWeather.css('div.symbol').text
    @currentTemperature = currentWeather.css('div.now-hero__next-hour-detail').first.text
    @currentPrecipitation = currentWeather.css('span.precipitation').text.gsub("Precipitation ", "")
    @currentWind = currentWeather.css('span.wind__value').text
    @currentWindDirection = currentWeather.css('div.wind-arrow__arrow-container').text

end

file = File.read('./weather.json')
data_hash = JSON.parse(file)

scraper

data_hash['time'] = @currentTime
data_hash['icon'] = @currentWeatherIcon
data_hash['temperature'] = @currentTemperature
data_hash['precipitation'] = @currentPrecipitation
data_hash['wind']['speed'] = @currentWind
data_hash['wind']['direction'] = @currentWindDirection

File.write('./weather.json', JSON.pretty_generate(data_hash))

File: scraper.js
// Ensure we use puppeteer
const puppeteer = require("puppeteer");

// Ensure we have Firebase too
const admin = require("firebase-admin");
const serviceAccount = require("./weatherExtensionServiceAccountKey.json");

//initialize admin SDK using serciceAcountKey
admin.initializeApp({
    credential: admin.credential.cert(serviceAccount),
});
const db = admin.firestore();

(async () => {
    
  const browser = await puppeteer.launch();
  const page = await browser.newPage();



  // now specify the page we want to scrape
  await page.goto(
    "https://www.yr.no/en/forecast/daily-table/2-2192362/New%20Zealand/Canterbury/Christchurch%20City/Christchurch"
  );

  const currentLocation = await page.evaluate(
    () => document.querySelector(".page-header__location-name").textContent.replace(" ", "")
  );
  const currentWeatherIcon = await page.evaluate(
    () => document.querySelector(".symbol").textContent
  );
  const currentTemperature = await page.evaluate(() =>
    document
      .querySelector(".temperature")
      .textContent.replace("Temperature", "")
  );
  const currentFeelsLikeTemperature = await page.evaluate(
    () => document.querySelector(".feels-like-text").textContent
  );
  const currentPrecipitation = await page.evaluate(
    () => document.querySelector(".precipitation").textContent
  );

  const currentWindSpeed = await page.evaluate(
    () => document.querySelector(".wind__value").textContent
  );
  const currentWindDirection = await page.evaluate(() =>
    document.querySelector(".wind .nrk-sr").textContent.substring(6)
  );

  // Write data to Firestore
  function writeCurrentWeather() {
    //return a promise since we'll imitating an API call
    return new Promise(function (resolve, reject) {
      resolve({
        location: currentLocation,
        icon: currentWeatherIcon,
        temperature: currentTemperature,
        feelsLikeTemperature: currentFeelsLikeTemperature,
        precipitation: currentPrecipitation,
        windSpeed: currentWindSpeed,
        windDirection: currentWindDirection
      });
    });
  }
  writeCurrentWeather().then((result) => {
    console.log(result);
    const obj = result;
    const weatherData = {
      location: obj.location,
      icon: obj.icon,
      temperature: obj.temperature,
      feelsLikeTemperature: obj.feelsLikeTemperature,
      precipitation: obj.precipitation,
      windSpeed: obj.windSpeed,
      windDirection: obj.windDirection
    };
    return db
      .collection('currentWeather')
      .doc('Christchurch')
      .set(weatherData)
      .then(() => console.log("Current weather forecast for " + currentLocation + " written to database"));
  });

  await browser.close();
})();
